src/Microsoft.Health.Fhir.SqlServer/Features/Storage/TvpRowGeneration/Merge/CompartmentAssignmentListRowGenerator.cs

```diff
@@ -0,0 +1,137 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using EnsureThat;
+using Microsoft.Health.Fhir.Core.Features.Persistence;
+using Microsoft.Health.Fhir.Core.Models;
+using Microsoft.Health.Fhir.SqlServer.Features.Schema.Model;
+using Microsoft.Health.SqlServer.Features.Schema.Model;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Storage.TvpRowGeneration
+{
+    internal class CompartmentAssignmentListRowGenerator : ITableValuedParameterRowGenerator<IReadOnlyList<MergeResourceWrapper>, CompartmentAssignmentListRow>
+    {
+        private readonly ISqlServerFhirModel _model;
+        private readonly SearchParameterToSearchValueTypeMap _searchParameterTypeMap;
+        private bool _initialized;
+        private byte _patientCompartmentId;
+        private byte _encounterCompartmentId;
+        private byte _relatedPersonCompartmentId;
+        private byte _practitionerCompartmentId;
+        private byte _deviceCompartmentId;
+
+        public CompartmentAssignmentListRowGenerator(ISqlServerFhirModel model, SearchParameterToSearchValueTypeMap searchParameterTypeMap)
+        {
+            EnsureArg.IsNotNull(model, nameof(model));
+            EnsureArg.IsNotNull(searchParameterTypeMap, nameof(searchParameterTypeMap));
+
+            _model = model;
+            _searchParameterTypeMap = searchParameterTypeMap;
+        }
+
+        public IEnumerable<CompartmentAssignmentListRow> GenerateRows(IReadOnlyList<MergeResourceWrapper> resources)
+        {
+            EnsureInitialized();
```

Comment:
Low priority: Should we add a validation to avoid null references of "resources"?

Reply:
Again, it is copy/past from current code. I don't see the need to add checks. I can try making classes internal, do it is obvious that they rely on external validation.